File: neetcode-150/0200-number-of-islands/Solution.java
class Solution {

    private int m, n;
    
    public int numIslands(char[][] grid) {
        int count = 0;
        n = grid.length;
        m = grid[0].length;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[j][i] == '1') {
                    count++;
                    dfs(grid, i, j);
                }
            }
        }
        
        return count;
    }
    
    public void dfs(char[][] grid, int i, int j) {
        if (i < 0 || j < 0 || i >= m || j >= n || grid[j][i] == '0') return;
        grid[j][i] = '0';
        dfs(grid, i + 1, j);
        dfs(grid, i - 1, j);
        dfs(grid, i, j + 1);
        dfs(grid, i, j - 1);
    }
}

File: neetcode-150/0200-number-of-islands/solution.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        n = len(grid)
        m = len(grid[0])
        
        def dfs(grid, i, j) :
            if i < 0 or j < 0 or i >= m or j >= n or grid[j][i] == '0':
                return

            grid[j][i] = '0';
            dfs(grid, i + 1, j);
            dfs(grid, i - 1, j);
            dfs(grid, i, j + 1);
            dfs(grid, i, j - 1);
        
        count = 0
        for j, row in enumerate(grid):
            for i, num in enumerate(row):
                if num == "1":
                    dfs(grid, i, j)
                    count += 1
        return count